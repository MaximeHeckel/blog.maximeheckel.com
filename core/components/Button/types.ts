import { ForwardRefComponent, HTMLMotionProps } from 'framer-motion';

export type MainButtonVariant = 'primary' | 'secondary';
export type IconButtonVariant = 'icon';

interface BaseButtonProps extends React.HTMLAttributes<HTMLButtonElement> {
  as?: ForwardRefComponent<HTMLButtonElement, HTMLMotionProps<'button'>>;
  type?: 'button' | 'reset' | 'submit';
}

interface MainButtonProps extends BaseButtonProps {
  variant: MainButtonVariant;
  icon?: never;
  startIcon?: React.ReactNode;
  endIcon?: React.ReactNode;
}

interface IconButtonProps extends BaseButtonProps {
  variant: IconButtonVariant;
  icon: React.ReactNode | HTMLOrSVGElement;
  startIcon?: never;
  endIcon?: never;
}

export type ButtonProps<T> = (MainButtonProps | IconButtonProps) & T;
