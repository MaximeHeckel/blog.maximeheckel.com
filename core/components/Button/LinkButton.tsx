import { motion } from 'framer-motion';
import styled from '@emotion/styled';

const LinkButton = styled(motion.button)`
  border-radius: var(--border-radius-1);
  width: 40px;
  height: 40px;
  cursor: pointer;
  border: none;
  background-color: transparent;
  transition: 0.4s;
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 0px;
  color: var(--maximeheckel-colors-typeface-tertiary);

  > * {
    margin-left: auto;
    margin-right: auto;
  }

  &:focus:not(:focus-visible) {
    outline: 0;
  }

  &:focus-visible {
    outline: 2px solid var(--maximeheckel-colors-brand);
    background-color: var(--maximeheckel-colors-foreground);
    box-shadow: var(--maximeheckel-shadow-2);
  }

  &:hover {
    background-color: var(--maximeheckel-colors-foreground);
    box-shadow: var(--maximeheckel-shadow-2);
  }
`;

export { LinkButton };
