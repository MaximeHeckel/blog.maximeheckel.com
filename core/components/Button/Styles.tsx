import { styled } from 'lib/stitches.config';

export const StyledButton = styled('button', {
  WebkitAppearance: 'none',
  WebkitTapHighlightColor: 'transparent',
  WebkitUserSelect: 'none',
  MozUserSelect: 'none',
  MsUserSelect: 'none',
  userSelect: 'none',
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  flexShrink: '0',
  outline: 'none',
  cursor: 'pointer',
  border: '0',
  font: 'inherit',
  /* Constant properties */
  fontSize: 'var(--font-size-2)',
  fontWeight: 'var(--font-weight-3)',
  height: '44px',
  width: 'max-content',
  padding: '11px 16px',
  transition: 'background 0.2s, transform 0.2s, color 0.2s, box-shadow 0.3s',
  borderRadius: 'var(--border-radius-1)',
  /* Computed properties */
  background: 'var(--background, white)',
  color: 'var(--color, black)',
  transform: 'scale(var(--button-scale, 1)) translateZ(0)',
  boxShadow: 'var(--shadow, none)',
  opacity: 'var(--opacity, 1)',
  '--shadow-hover-primary':
    '0 2px 40px -4px var(--maximeheckel-form-input-focus)',

  '&:active': {
    '--button-scale': '0.95',
  },
  '&:disabled': {
    cursor: 'not-allowed',
  },

  variants: {
    variant: {
      primary: {
        '--background': 'var(--maximeheckel-colors-brand)',
        '--color': 'hsl(var(--palette-gray-00))',
        '&:disabled': {
          '--background': 'var(--maximeheckel-form-input-disabled)',
          '--color': 'var(--maximeheckel-colors-typeface-tertiary)',
        },

        '&:hover': {
          '&:not(:disabled)': {
            '--shadow': 'var(--shadow-hover-primary)',
          },
        },

        '&:focus-visible': {
          '--shadow': 'var(--shadow-hover-primary)',
        },
      },
      secondary: {
        '--background': 'var(--maximeheckel-colors-emphasis)',
        '--color': 'var(--maximeheckel-colors-brand)',
        '&:disabled': {
          '--background': 'var(--maximeheckel-form-input-disabled)',
          '--color': 'var(--maximeheckel-colors-typeface-tertiary)',
        },

        '&:hover': {
          '&:not(:disabled)': {
            '--shadow': 'var(--shadow-hover-primary)',
          },
        },

        '&:focus-visible': {
          '--shadow': 'var(--shadow-hover-primary)',
        },
      },
    },
  },
});

export const StyledIconButton = styled('button', {
  WebkitAppearance: 'none',
  WebkitTapHighlightColor: 'transparent',
  WebkitUserSelect: 'none',
  MozUserSelect: 'none',
  MsUserSelect: 'none',
  userSelect: 'none',
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  flexShrink: '0',
  outline: 'none',
  cursor: 'pointer',
  border: '0',
  /* Constant properties */
  width: '44px',
  height: '44px',
  background: 'transparent',
  transition: 'color 0.3s ease, transform 0.3s ease',
  borderRadius: 'var(--border-radius-1)',
  color: 'var(--color, var(--maximeheckel-colors-typeface-secondary))',
  transform: 'scale(var(--button-content-scale, 1)) translateZ(0)',
  '--shadow-hover-primary':
    '0 2px 40px -4px var(--maximeheckel-form-input-focus)',

  '&::after': {
    zIndex: '0',
    position: 'absolute',
    content: "''",
    display: 'block',
    width: '100%',
    height: '100%',
    borderRadius: 'var(--corner, var(--border-radius-1))',
    transition:
      'box-shadow 0.3s ease, border-color 0.2s, background 0.3s ease,\n      transform 0.3s cubic-bezier(0.34, 1.56, 0.64, 1)',
    background: 'var(--background, var(--maximeheckel-colors-foreground))',
    transform: 'scale(var(--button-background-scale, 1)) translateZ(0)',
    border:
      '2px solid var(--border-color, var(--maximeheckel-colors-foreground))',
    boxShadow: 'var(--shadow, none)',
  },

  '&:disabled': {
    cursor: 'not-allowed',
    '--background': 'var(--maximeheckel-form-input-disabled)',
    '--color': 'var(--maximeheckel-colors-typeface-tertiary)',
  },

  '&:hover': {
    '&:not(:disabled)': {
      '--border-color': 'var(--maximeheckel-colors-brand)',
      '--color': 'var(--maximeheckel-colors-brand)',
      '--corner': 'calc(var(--border-radius-1) + 2px)',
      '--button-background-scale': '0.92',
      '--shadow': 'var(--shadow-hover-primary)',
    },
  },

  '&:focus-visible': {
    '--border-color': 'var(--maximeheckel-colors-brand)',
    '--color': 'var(--maximeheckel-colors-brand)',
    '--corner': 'calc(var(--border-radius-1) + 2px)',
    '--button-background-scale': 0.92,
    '--shadow': 'var(--shadow-hover-primary)',
  },

  '&:active': {
    '--button-content-scale': '0.95',
  },
});
