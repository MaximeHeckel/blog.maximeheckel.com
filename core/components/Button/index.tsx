export { default } from './Button';
export type { ButtonProps } from './types';
