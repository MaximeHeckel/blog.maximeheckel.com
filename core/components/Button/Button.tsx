import React from 'react';
import Flex from '../Flex';
import { StyledButton, StyledIconButton } from './Styles';
import { ButtonProps } from './types';

const Button = <T extends object>(
  props: ButtonProps<T>,
  ref: React.ForwardedRef<HTMLButtonElement>
) => {
  const {
    variant = 'primary',
    children,
    icon,
    startIcon,
    endIcon,
    as: Component,
    ...rest
  } = props;

  if (variant === 'icon') {
    return (
      <StyledIconButton variant={variant} as={Component} ref={ref} {...rest}>
        <Flex
          css={{
            zIndex: 1,
          }}
        >
          {icon}
        </Flex>
      </StyledIconButton>
    );
  }

  return (
    <StyledButton variant={variant} as={Component} ref={ref} {...rest}>
      {startIcon ? (
        <Flex
          css={{
            marginRight: '8px',
          }}
        >
          {startIcon}
        </Flex>
      ) : null}
      {children}
      {endIcon ? (
        <Flex
          css={{
            marginLeft: '8px',
          }}
        >
          {endIcon}
        </Flex>
      ) : null}
    </StyledButton>
  );
};
Button.displayName = 'Button';

const ForwardedButton = React.forwardRef(Button);

export const WrappedButton = <T,>({
  ref,
  ...rest
}: ButtonProps<T> & { ref?: React.Ref<HTMLButtonElement> }) => (
  <ForwardedButton {...rest} ref={ref} />
);

export default WrappedButton;
