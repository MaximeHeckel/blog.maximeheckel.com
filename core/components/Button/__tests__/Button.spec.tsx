import { render } from '@testing-library/react';
import React from 'react';
import Button from '../';

describe('Button', () => {
  it('renders a basic button', () => {
    const { container, getByText } = render(
      <Button variant="primary">Test</Button>
    );
    expect(container.querySelector('button')).toBeInTheDocument();
    expect(container.querySelector('button')).toMatchSnapshot();
    expect(getByText('Test')).toBeInTheDocument();
  });

  it('renders a button with a startIcon', () => {
    const { container, getByTestId, getByText } = render(
      <Button variant="primary" startIcon={<svg data-testid="icon"></svg>}>
        Test
      </Button>
    );
    expect(container.querySelector('button')).toBeInTheDocument();
    expect(container.querySelector('button')).toMatchSnapshot();
    expect(getByTestId('icon')).toBeInTheDocument();
    expect(getByText('Test')).toBeInTheDocument();
  });

  it('renders a button with an endIcon', () => {
    const { container, getByTestId, getByText } = render(
      <Button variant="primary" endIcon={<svg data-testid="icon"></svg>}>
        Test
      </Button>
    );
    expect(container.querySelector('button')).toBeInTheDocument();
    expect(container.querySelector('button')).toMatchSnapshot();
    expect(getByTestId('icon')).toBeInTheDocument();
    expect(getByText('Test')).toBeInTheDocument();
  });

  it('renders an icon button', () => {
    const { container, getByTestId, debug } = render(
      <Button variant="icon" icon={<svg data-testid="icon"></svg>} />
    );
    expect(container.querySelector('button')).toBeInTheDocument();
    expect(container.querySelector('button')).toMatchSnapshot();
    expect(getByTestId('icon')).toBeInTheDocument();
  });
});
