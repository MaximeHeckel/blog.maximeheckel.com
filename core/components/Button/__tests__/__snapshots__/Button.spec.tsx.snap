// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders a basic button 1`] = `
<button
  class="c-bVoOVF c-bVoOVF-kxlbib-variant-primary"
>
  Test
</button>
`;

exports[`Button renders a button with a startIcon 1`] = `
<button
  class="c-bVoOVF c-bVoOVF-kxlbib-variant-primary"
>
  <span
    style="margin-right: 8px;"
  >
    <svg
      data-testid="icon"
    />
  </span>
  Test
</button>
`;

exports[`Button renders a button with an endIcon 1`] = `
<button
  class="c-bVoOVF c-bVoOVF-kxlbib-variant-primary"
>
  Test
  <span
    style="margin-left: 8px;"
  >
    <svg
      data-testid="icon"
    />
  </span>
</button>
`;

exports[`Button renders an icon button 1`] = `
<button
  class="c-bhXMVC"
  variant="icon"
>
  <span
    style="z-index: 1; display: flex;"
  >
    <svg
      data-testid="icon"
    />
  </span>
</button>
`;
